File: Excel/621 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/621 Palindromic Dates in 2025.xlsx"
input = read_excel(path, range = "A2:A14")
test  = read_excel(path, range = "A2:B14")

result = seq.Date(as.Date("2025-01-01"), as.Date("2025-12-31"), by = "day") %>%
  as_tibble() %>%
  separate(value, c("year", "month", "day"), sep = "-", remove = F) %>%
  mutate(year = as.numeric(year) - 2000,
         month = as.numeric(month),
         day = as.numeric(day),
         value = as.POSIXct(value)) %>%
  mutate(MDY = paste(month, day, year, sep = ""),
         DMY = paste(day, month, year, sep = ""),
         YMD = paste(year, month, day, sep = "")) %>%
  select(Date = value, MDY, DMY, YMD) %>%
  pivot_longer(cols = -Date, names_to = "Format", values_to = "Value") %>%
  mutate(Rev = map_chr(Value, ~str_c(rev(str_split(.x, "")[[1]]), collapse = ""))) %>%
  filter(Value == Rev) %>%
  summarise(Format = str_c(Format, collapse = ", "), .by = Date)

all.equal(result, test, check.attributes = F)
#> [1] TRUE

File: Excel/600-699/621/621 Challenge.py
import pandas as pd

path = "621 Palindromic Dates in 2025.xlsx"
input = pd.read_excel(path, usecols="A", skiprows=1, nrows=12)
test = pd.read_excel(path, usecols="A:B", skiprows=1, nrows=13)

date_range = pd.date_range(start="2025-01-01", end="2025-12-31")

df = pd.DataFrame({'Date': date_range})
df['year'], df['month'], df['day'] = df['Date'].dt.year - 2000, df['Date'].dt.month, df['Date'].dt.day

df['MDY'] = df['month'].astype(str) + df['day'].astype(str) + df['year'].astype(str)
df['DMY'] = df['day'].astype(str) + df['month'].astype(str) + df['year'].astype(str)
df['YMD'] = df['year'].astype(str) + df['month'].astype(str) + df['day'].astype(str)

df_long = df.melt(id_vars=['Date'], value_vars=['MDY', 'DMY', 'YMD'], var_name='Format', value_name='Value')

df_long['Rev'] = df_long['Value'].apply(lambda x: x[::-1])
result = df_long[df_long['Value'] == df_long['Rev']].groupby('Date')['Format'].apply(lambda x: ', '.join(x)).reset_index()

print(result.equals(test)) # True